source/detail/time.hpp

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2024 xlnt-community
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+//
+// @license: http://www.opensource.org/licenses/mit-license.php
+// @author: see AUTHORS file
+
+#pragma once
+
+#include <ctime>
+
+namespace xlnt {
+namespace detail {
+
+inline std::tm localtime_safe(std::time_t raw_time)
+{
+    std::tm result{};
+
+#ifdef _MSC_VER
+    localtime_s(&result, &raw_time);
+#elif _POSIX_C_SOURCE >= 1 || _XOPEN_SOURCE || _BSD_SOURCE || _SVID_SOURCE || _POSIX_SOURCE
```

Comment:
What is the reason for using those constants? Shouldn't we use the `defined` operator in these checks? In [source/utils/path.cpp](https://github.com/xlnt-community/xlnt/blob/0b772584c11fc342dba35401ac53e85bf41cb654/source/utils/path.cpp#L32) `__linux` and `__APPLE__` are used. Should we use the same here?

Reply:
I took the feature test macros from here: https://linux.die.net/man/3/localtime_r

The thing is, in my experience, the more you go into platform-specific functions, the harder it gets - and, unfortunately, this often goes beyond differentiating between Windows, Linux and macOS. Because I just looked over miniz yesterday - look what they are doing in order to support all the different platforms: https://github.com/richgel999/miniz/blob/master/miniz_zip.c#L104

My point is: I want to use the new code to improve code quality and safety - however, due to the code being less portable, we need to be much more careful to still compile on different platforms. Since I did not find any better documentation on the feature test macros, I think the current solution is the best way to go to in order to ensure maximum portability :slightly_smiling_face: 

Some examples I heard of but did not experience them myself: Android is Linux, but Android still has its own SDK which might be different in some ways. Also, there are some differences between Clang and Apple Clang, so again, just testing for one might lead to wrong results. In the end, the only real way to really make sure it works on as many platforms as possible is to let the community test it, but at least we can try our best to write portable code :slightly_smiling_face: